### PR TITLE
Add scraper implementation

### PR DESCRIPTION
**Pull request by MNV**

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+venv
+logs
+Makefile
+!.gitignore
+!.dockerignore
```
Comment (joein) [inline .dockerignore]: are !.gitignore and !.dockerignore actually required?

File: black.toml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+[tool.black]
+line-length = 120
```
Comment (joein) [inline black.toml]: let's use at most 99

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+[tool.poetry.dependencies]
+python = "^3.10"
+# парсинг
+Scrapy = "^2.7.0"
+scrapy-user-agents = "^0.1.1"
+# работа с MongoDB
+pymongo = {version = "^4.3.2", extras = ["srv"]}
+
+[tool.poetry.dev-dependencies]
```
Comment (joein) [inline pyproject.toml]: https://python-poetry.org/docs/master/managing-dependencies/#dependency-groups

File: src/vacancies/vacancies/items.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    """
+    Информация о ближайших станциях метро.
+    """
+
+    station_name = Field()
+    line_name = Field()
+    station_id = Field()
+    line_id = Field()
+    lat = Field()
+    lng = Field()
```
Comment (joein) [inline src/vacancies/vacancies/items.py]: Looks exactly the same as `Metro`

File: src/vacancies/vacancies/spiders/vacancies.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    def parse(self, response: TextResponse, **kwargs: Any) -> Optional[Generator]:
+        """
+        Парсинг данных о регионах.
+
+        :param response: Ответ от сервера.
+        :param kwargs:
+        :return:
+        """
+
+        json_res = json.loads(response.body)
```
Comment (joein) [inline src/vacancies/vacancies/spiders/vacancies.py]: How scrapy handle all the exceptions which may occur during parsing?
What happens if host drops the connection or return something different from 200 status code?

According to the docs `response.body` can contain empty byte-string (not sure if it is the case here), but `json` fails to deserialize it properly, maybe it should be handled?

File: src/vacancies/vacancies/spiders/vacancies.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+    def parse_pages(self, response: TextResponse, **kwargs: Any) -> Optional[Generator]:
+        """
+        Парсинг списка вакансий в указанном регионе с пагинацией.
+
+        :param response: Ответ от сервера.
+        :param kwargs:
+        :return:
+        """
+
+        json_res = json.loads(response.body)
```
Comment (joein) [inline src/vacancies/vacancies/spiders/vacancies.py]: same question as in the previous method, can we get anything non-deserializable here and what will happen in such a situation?

File: src/vacancies/vacancies/spiders/vacancies.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        if not json_res:
+            return None
+
+        vacancy = VacancyItem()
```
Comment (joein) [inline src/vacancies/vacancies/spiders/vacancies.py:146]: reduce number of lines via a similar constructions where it is possible

```Python
keys = {"id", "premium", "relations", "name", "insider_interview", "response_letter_required", "salary", ...}
vacancy = {key: json_res[key] for key in keys}
```

**Review by joein — CHANGES_REQUESTED** [review]
avoid creating huge pull requests, split them into several logically related ones

File: .dockerignore (hunk 1 above)
Comment (MNV) [reply .dockerignore]: They are not required. I've updated the file using best practice examples

File: black.toml (hunk 2 above)
Comment (MNV) [reply black.toml]: Done. Code has been reformatted

File: pyproject.toml (hunk 3 above)
Comment (MNV) [reply pyproject.toml]: Added dependency groups splitting the reqs

File: src/vacancies/vacancies/items.py (hunk 4 above)
Comment (MNV) [reply src/vacancies/vacancies/items.py]: Right. I've removed the extra class

File: src/vacancies/vacancies/spiders/vacancies.py (hunk 5 above)
Comment (MNV) [reply src/vacancies/vacancies/spiders/vacancies.py]: 1. Scrapy middleware like `scrapy.spidermiddlewares.httperror` would automatically ignore the "not OK" status-code with the output:
```
[scrapy.spidermiddlewares.httperror] INFO: Ignoring response <400 https://URL Ardess>: HTTP status code is not handled or not allowed
```

2. For `json.loads` case I've added special handling.

File: src/vacancies/vacancies/spiders/vacancies.py (hunk 6 above)
Comment (MNV) [reply src/vacancies/vacancies/spiders/vacancies.py]: For `json.loads` case I've added special handling.

File: src/vacancies/vacancies/spiders/vacancies.py (hunk 7 above)
Comment (MNV) [reply src/vacancies/vacancies/spiders/vacancies.py:146]: Fixed using this suggestion

File: pyproject.toml [hunk 8]
```diff
@@ -12,18 +12,22 @@ Scrapy = "^2.7.0"
 scrapy-user-agents = "^0.1.1"
 # работа с MongoDB
 pymongo = {version = "^4.3.2", extras = ["srv"]}
+# валидация данных
+pydantic = "^1.10.2"
 
-[tool.poetry.dev-dependencies]
 # документация
+[tool.poetry.group.documentation.dependencies]
```
Comment (joein) [inline pyproject.toml:22]: make documentation, tests and linters groups optional

File: src/vacancies/vacancies/spiders/vacancies.py (hunk 7 above)
Comment (joein) [reply src/vacancies/vacancies/spiders/vacancies.py:146]: same pattern can be applied to fields like "employer", but it is up to you

**Review by joein — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 8 above)
Comment (MNV) [reply pyproject.toml:22]: Done

File: src/vacancies/vacancies/spiders/vacancies.py (hunk 7 above)
Comment (MNV) [reply src/vacancies/vacancies/spiders/vacancies.py:146]: These fields require more complex processing. I think appropriate refactoring can be made later

**Review by joein — APPROVED** [review]

**Pull request merged**